### DISCUSSION TITLE
pyenv support

### DISCUSSION THREAD

**Discussion by jefftriplett** [brettcannon/python-launcher#107] [Ideas]
I am not sure how far you want to deep dive with having pyenv support, but I noticed this behavior today: 

```shell
❯ python --version
Python 3.9.5

❯ py --version
Python 3.10.0a4+
```

Basically, `pyenv global` lets you set up a bunch of global python versions including which version is your default aka resolves first when you run `python` or `python3`.

At first glance, this might not seem important, but VS Code and a few other apps, do not like it if you have Python 3.6 installed, but not linked to a global version like `python3.6`. 

```shell
❯ pyenv global
3.9.5
3.8.10
3.7.10
3.10-dev
3.6.13
3.5.10
2.7.18

# this order is installed here:
❯ cat ~/.pyenv/version
3.9.5
3.8.10
3.7.10
3.10-dev
3.6.13
3.5.10
2.7.18
```

I would expect `py --version` to give me `Python 3.9.5` instead of `Python 3.10.0a4+` because other apps seem to just work with the default `python` or `python3` version that I have setup. 

`py -3.9 --version` works but is less tab-completable than `py`. 

PS: I don't think this is a deal-breaker, but I suspect it'll be a common issue now that more and more tutorials and guides are starting to recommend that people use pyenv over homebrew or other python versions.

**Comment by brettcannon** [comment]
You can set `PY_PYTHON` to the Python version you want the default to point to, e.g. `PY_PYTHON=3.9` would mean your 3.9.5 install would take precedence over your 3.10-dev install.

But I assume you still want explicit pyenv support? Or does that environment variable solve this for you?

**Reply by jefftriplett** [reply]
@brettcannon, I assume that using `PY_PYTHON` globally would mean that I'd have to set it on a venv (or virtualenv) level to override it? 

If py ignores `PY_PYTHON` once you are in an activated venv/virtualenv, this is a good workaround. (I think this is already the default from my 15 minutes of playing around.) 

If py does not ignore `PY_PYTHON`, then that might be a pain because I wouldn't want one global PY_PYTHON change to break any existing venvs set up with an older Python version. 

Ideally, it would be nice to respect pyenv's global pick by default, but I understand that might be out of scope for py. (I suspect it's going to come up again.)

**Reply by brettcannon** [reply]
The flowchart at https://github.com/brettcannon/python-launcher#documentation should make the priority clear (otherwise I need to redo the flowchart 😉 ). You can also run the Launcher with logging turned on if you want more details.

But yes, the Launcher only uses `PY_PYTHON` if it couldn't find a fitting virtual environment.

And supporting pyenv's personal configuration file format does feel out of scope since it isn't standardized in any way.

**Reply by brettcannon** [reply]
I think you can get what you want with:

```
PY_PYTHON=head --lines 1 ​~​/.pyenv/version
```

Does that work?

**Reply by jefftriplett** [reply]
In theory it should work, but it still hates me ;) 

```
❯ PY_PYTHON=3.9.5 && py --version
Python 3.10.0a4+

❯ echo $PY_PYTHON
3.9.5

❯ py --version
Python 3.10.0a4+
```

**Reply by brettcannon** [reply]
Ah, I realize my mistake. `PY_PYTHON` only takes the `major.minor` format, so `3.9.5` won't parse. That should show up in the log.

I've opened https://github.com/brettcannon/python-launcher/issues/108 to track making sure this kind of failure has some possibility of being user-identifiable.

**Reply by benspaulding** [reply]
I came here for a similar reason as Jeff, a little tripped up at `py` giving me a different version than `python` when I have `pyenv` installed. In my case it was because `pyenv` respects `.python-version` files.

I tried to get it to work with a `.python-version` file and hit the same issue.

```fish
> cat .python-version; and set -gx PY_PYTHON (head -n1 .python-version); and echo $PY_PYTHON; and py --version
3.7.10
3.7.10
Python 3.9.5
```

Is making `PY_PYTHON` accept `major.minor.patch` a possibility, or does that go against some of `py`’s design decisions?

This is an awesome, project — thanks for your efforts!

**Reply by treyhunner** [reply]
I put this in my ~/.zshrc` and it worked:

```sh
export PY_PYTHON=$(head --lines 1 ~/.python-version | cut --delimiter=. --fields=1,2)
```

Before I saw this:

```bash
$ py --version
Python 3.10.0a6+
```

And now:

```bash
$ py --version
Python 3.9.1
```

**Reply by brettcannon** [reply]
@benspaulding I will have to think about whether I want to be lenient on `PY_PYTHON` with a patch version, as I also don't want people to think that by specifying a patch version they will actually get that specific version as they won't.

@treyhunner Thanks for that! Do you want to send me a PR to add that to the FAQ in the README (with a note that the syntax is for bash/zsh and for fish it's `set -g -x (head --lines 1 ~/.python-version | cut --delimiter=. --fields=1,2)`)? If not, I assume you don't mind if I copy that into there?

**Reply by treyhunner** [reply]
@brettcannon PR opened #111 👍

Feel free to reword as desired

**Reply by brettcannon** [reply]
If anyone is interested, https://github.com/brettcannon/python-launcher/issues/108 is looking for feedback on how to surface any parsing issues with the environment variables.

**Comment by brettcannon** [comment]
This is now covered in the README thanks to https://github.com/brettcannon/python-launcher/commit/c9e92a5db5b3993031ccff05660e5bfe09061951! Thanks everyone for helping out to resolve this!
